packages/connection-form/src/utils/csfle-kms-fields.ts

```diff
@@ -0,0 +1,172 @@
+import type { AutoEncryptionOptions } from 'mongodb';
+import type { ConnectionFormError } from './validation';
+import { errorMessageByFieldName, fieldNameHasError } from './validation';
+
+type KeysOfUnion<T> = T extends T ? keyof T : never;
+export type KMSProviders = NonNullable<AutoEncryptionOptions['kmsProviders']>;
+export type KMSProviderName = keyof KMSProviders;
+export type KMSOption<KMSProvider extends KMSProviderName> = KeysOfUnion<
+  NonNullable<KMSProviders[KMSProvider]>
+>;
+
+export interface KMSField<KMSProvider extends KMSProviderName> {
+  name: KMSOption<KMSProvider>;
+  label: string;
+  type: 'password' | 'text' | 'textarea';
+  optional: boolean;
+  value: (autoEncryption: AutoEncryptionOptions) => string;
+  errorMessage?: (errors: ConnectionFormError[]) => string | undefined;
+  state:
+    | 'error'
+    | 'none'
+    | ((errors: ConnectionFormError[]) => 'error' | 'none');
+  description?: string;
+}
+
+const GCPFields: KMSField<'gcp'>[] = [
+  {
+    name: 'email',
+    label: 'Service Account E-Mail',
+    type: 'text',
+    optional: false,
+    value: (autoEncryption) => autoEncryption?.kmsProviders?.gcp?.email ?? '',
+    state: 'none',
+    description: 'The service account email to authenticate.',
+  },
+  {
+    name: 'privateKey',
+    label: 'Private Key',
+    type: 'textarea',
+    optional: false,
+    value: (autoEncryption) =>
+      autoEncryption?.kmsProviders?.gcp?.privateKey?.toString('base64') ?? '',
+    state: 'none',
+    description: 'A base64-encoded PKCS#8 private key.',
+  },
+  {
+    name: 'endpoint',
+    label: 'Endpoint',
+    type: 'text',
+    optional: true,
+    value: (autoEncryption) =>
+      autoEncryption?.kmsProviders?.gcp?.endpoint ?? '',
+    state: 'none',
+    description: 'A host with optional port.',
```

Comment:
Maybe
```suggestion
    description: 'A host with an optional port.',
```
reads better?